kart/spatial_tree.py

```diff
@@ -0,0 +1,430 @@
+        self.blob_cells = Table(
+            "blob_cells",
+            self.sqlalchemy_metadata,
+            # Reference to blobs.rowid.
+            Column(
+                "blob_rowid",
+                Integer,
+                ForeignKey("blobs.rowid"),
+                nullable=False,
+                primary_key=True,
+            ),
+            # S2 cell token eg "6d6dd90351b31cbf".
+            # To locate an S2 cell by token, see https://s2.sidewalklabs.com/regioncoverer/
+            Column(
+                "cell_token",
+                Text,
+                nullable=False,
+                primary_key=True,
+            ),
+        )
+
+
+SpatialTreeTables.copy_tables_to_class()
+
+
+def drop_tables(sess):
+    sess.execute("DROP TABLE IF EXISTS blob_cells;")
+    sess.execute("DROP TABLE IF EXISTS blobs;")
+
+
+def iter_feature_oids(repo, commit_spec):
+    cmd = _revlist_command(repo) + commit_spec
+    try:
+        p = subprocess.Popen(
+            cmd,
+            stdout=subprocess.PIPE,
+            encoding="utf8",
+            env=tool_environment(),
+        )
+        yield from _parse_revlist_output(p.stdout, r"feature/.+")
+    except subprocess.CalledProcessError as e:
+        raise SubprocessError(
+            f"There was a problem with git rev-list: {e}", called_process_error=e
+        )
+
+
+def update_spatial_tree(repo, commit_spec, verbosity=1, clear_existing=False):
+    """
+    Index the commits given in commit_spec, and write them to the s2_index.db repo file.
+
+    repo - the Kart repo containing the commits to index, and in which to write the index file.
+    commit_spec - a list of commits to index (ancestors of these are implicitly included).
+        Commits can be exluded by prefixing with '^' (ancestors of these are implicitly excluded).
+        (See git rev-list for the full list of possibilities for specifying commits).
+    verbosity - how much non-essential information to output.
+    clear_existing - when true, deletes any pre-existing data before re-indexing.
+    """
+    import pywraps2 as s2
+
+    crs_helper = CrsHelper(repo)
+    feature_oid_iter = iter_feature_oids(repo, commit_spec)
+
+    s2_coverer = s2.S2RegionCoverer()
+    s2_coverer.set_max_cells(S2_MAX_CELLS_INDEX)
+    s2_coverer.set_max_level(S2_MAX_LEVEL)
+
+    progress_every = None
+    if verbosity >= 1:
+        progress_every = max(100, 100_000 // (10 ** (verbosity - 1)))
+
+    db_path = repo.gitdir_file(KartRepoFiles.S2_INDEX)
+    engine = sqlite_engine(db_path)
+    with sessionmaker(bind=engine)() as sess:
+        if clear_existing:
+            drop_tables(sess)
+
+        SpatialTreeTables.create_all(sess)
+
+    click.echo(f"Indexing {' '.join(commit_spec)} ...")
+    t0 = time.monotonic()
+    i = 0
+
+    # Using sqlite directly here instead of sqlalchemy is about 10x faster.
+    # Possibly due to huge number of unbatched queries.
+    # TODO - investigate further.
```

Comment:
Yeah, this needs to perform well... can we TODO it now?

Using SQLAlchemy shouldn't be magnitudes slower.

Reply:
It's using sqlite directly right now instead of sqlalchemy so it's fast right now. But I can take some time and try and figure out what's making sqlalchemy slow (might be relevant to working copies also)